Group by key

Michaelliv opened the pull request.

Michaelliv committed:
Black

Michaelliv committed:
Changed _Batching implementation to support grouping by key

Michaelliv committed:
Added GroupByKey implementation

Michaelliv committed:
Added GroupByKey tests

Michaelliv committed:
Merge branch 'development' into batch-by-key

# Conflicts:
#	storey/__init__.py
#	storey/flow.py

Michaelliv committed:
Black formatter

Michaelliv committed:
fixed linting

gtopper commented on storey/steps/groupbykey.py:
```diff
@@ -0,0 +1,32 @@
+    :param max_events: Maximum number of events per emitted batch. Set to None to emit all events in one batch on flow termination.
+    :type max_events: int or None
+    :param timeout_secs: Maximum number of seconds to wait before a batch is emitted.
+    :type timeout_secs: int
```
No need for :type when using type annotations.

Michaelliv replied:
Thats new to me, I'll remove it

gtopper commented on storey/flow.py:
```diff
@@ -127,7 +147,7 @@ class _UnaryFunctionFlow(Flow):
         super().__init__(**kwargs)
         if not callable(fn):
-            raise TypeError(f'Expected a callable, got {type(fn)}')
+            raise TypeError(f"Expected a callable, got {type(fn)}")
```
Please don't...

Michaelliv committed:
Partially rolling back formatter

Michaelliv committed:
Removed typing documentation

talIguaz commented on storey/flow.py:
```diff
@@ -608,63 +611,98 @@ async def _handle_completed(self, event, response):
+
+    async def _emit_batch(self, batch_key: Optional[str] = None):
+        batch_to_emit = self._batch.pop(batch_key)
+        if len(batch_to_emit) > 0:
```
i think you will only have an entry in `self._batch` if there are items in `batch_to_emit` so it's a redundant check

Michaelliv replied:
Yeah just following previous logic. Will change that.

talIguaz commented on storey/flow.py:
```diff
@@ -608,63 +611,98 @@ async def _handle_completed(self, event, response):
+            await self._emit_batch(key)
+
+    def _init_timeout_task(self):
+        for key in self._batch.keys():
```
Aren't you creating a task per key here? we decided to only have one task for the "next" key to emit

Michaelliv replied:
There's a break right after it. could have wrapped `self._batch.keys()` with a list, then take the first element, but that wouldn't be less messy.

talIguaz replied:
oh right, why don't just do `key = next(iter(self._batch))`

Michaelliv replied:
Nice, will do

talIguaz commented on storey/flow.py:
```diff
@@ -608,63 +611,98 @@ async def _handle_completed(self, event, response):
+        if time_delta < self._timeout_secs:
+            await asyncio.sleep(self._timeout_secs - time_delta)
+        await self._emit_batch(key)
+        self._timeout_task = None
```
rather than setting "yourself" to None you can just sleep till the next key to emit as long as there are keys in the dict

Michaelliv replied:
Please elaborate

talIguaz commented on storey/steps/groupbykey.py:
```diff
@@ -0,0 +1,29 @@
+from storey.flow import _Batching
+
+
+class GroupByKey(_Batching):
```
In my opinion it's better to just extend this functionality to `Batch`, rather than a new Step. also it's not really `GroupByKey` more of `BatchByKey`

Michaelliv replied:
Then maybe `Batch` should change to `Group` with an optional key, maybe it should follow a more standard name for this type of operation. I've never called a batch function, but i did call GroupBy.

What do you think?

talIguaz replied:
GroupBy is sort of a finite/reduce operation it's grouping the entire dataset, while Batch/BatchByKey just batches a bunch of events based on some logic(max events/timeout) and emits them forward possible by key.

talIguaz commented on tests/test_group_by_key.py:
```diff
@@ -0,0 +1,106 @@
+    values_2 = [i for i in range(3)]
+    values_3 = [i for i in range(3)]
+    for i in range(3):
+        controller.emit({"value": values_1[i], "data": values_1[i]})
```
You are sending the same key sequentially.
better have tests that emit the keys not in order (like [key1, key1, key2, key3, key2, key1]) and test scenarios where you reach max events for a certain key and scenarios where you reach timeout for a specific key, etc..

Michaelliv replied:
No problem

Michaelliv committed:
Exposed group by key in Batch class, removed GroupByKey class.
Updated test to show actual grouping.

Michaelliv committed:
removed redundant init

Michaelliv committed:
_sleep_and_emit now takes responsibility to continue sleeping as long as self._batch contains event. Also, events with different keys will not cancel _sleep_and_emit when unnecessary.

Michaelliv committed:
flake

Michaelliv committed:
missing import in tests

Michaelliv committed:
Fixed issue with key extraction

Michaelliv committed:
Merge branch 'development' into batch-by-key

# Conflicts:
#	tests/test_flow.py

talIguaz commented on storey/flow.py:
```diff
@@ -666,24 +655,52 @@ async def _do(self, event):
+                await asyncio.sleep(self._timeout_secs - delta_seconds)
             await self._emit_batch(key)
 
+        self._timeout_task.cancel()
```
This **is** the `_timeout_task` so you don't need to cancel it i guess just reset the timeout_key and finish the function.

talIguaz commented on tests/test_flow.py:
```diff
@@ -2,12 +2,15 @@
-from storey import build_flow, Source, Map, Filter, FlatMap, Reduce, FlowError, MapWithState, ReadCSV, Complete, AsyncSource, Choice, \
-    Event, Batch, Table, WriteToCSV, DataframeSource, MapClass, JoinWithTable, ReduceToDataFrame, ToDataFrame, WriteToParquet, \
+from storey import build_flow, Source, Map, Filter, FlatMap, Reduce, FlowError, MapWithState, ReadCSV, Complete, \
+    AsyncSource, Choice, \
```
Weird line breaking

talIguaz commented on storey/flow.py:
```diff
@@ -608,73 +611,118 @@ async def _handle_completed(self, event, response):
 class Batch(_Batching):
     """
-    Batches events into lists of up to max_events events. Each emitted list contained max_events events, unless timeout_secs seconds
+    Batches events into lists of up to max_events events. Each emitted list  contained max_events events, unless timeout_secs seconds
```
extra space

talIguaz commented on storey/flow.py:
```diff
@@ -608,73 +611,118 @@ async def _handle_completed(self, event, response):
     :type max_events: int or None
     :param timeout_secs: Maximum number of seconds to wait before a batch is emitted.
     :type timeout_secs: int
+    :param group_by_key: If true, Batch will will group events by given key
```
No need for `type` in docstring

talIguaz replied:
double `will`

Michaelliv replied:
> No need for `type` in docstring

Habits

gtopper commented on storey/flow.py:
```diff
@@ -608,73 +611,118 @@ async def _handle_completed(self, event, response):
+
+        self._event_count: Dict[Optional[str], int] = defaultdict(int)
+        self._batch: Dict[Optional[str], List[Any]] = defaultdict(list)
+        self._batch_time: Dict[Optional[str], datetime] = {}
```
Does it make sense for the key to be `None`?

talIguaz replied:
In case it's not grouped by key

gtopper commented on storey/flow.py:
```diff
@@ -608,73 +611,118 @@ async def _handle_completed(self, event, response):
 
-            await self._emit(batch_to_emit, batch_time)
+    async def _emit_all(self):
+        batch_keys = list(self._batch.keys())
+        for batch_key in batch_keys:
```
Just
```
for batch_key in self._batch:
```
does the trick without creating an intermediate list.

Michaelliv replied:
Can't do that since `self._emit_batch(batch_key)` changes `self._batch`'s size

gtopper commented on storey/flow.py:
```diff
@@ -608,73 +611,118 @@ async def _handle_completed(self, event, response):
+
         if self._timeout_secs is not None and self._timeout_secs <= 0:
-            raise ValueError('Batch timeout cannot be 0 or negative')
+            raise ValueError("Batch timeout cannot be 0 or negative")
```
Please leave as single quotes.

gtopper commented on storey/flow.py:
```diff
@@ -608,73 +611,118 @@ async def _handle_completed(self, event, response):
+            max_events: Optional[int] = None,
+            timeout_secs: Optional[int] = None,
+            group_by_key: bool = False,
+            key: Optional[Union[str, callable]] = None,
```
`key_field` or `key_extractor`, really. Do we need `group_by_key` at all? Can't it just be implied when key field is provided?

talIguaz replied:
No, because if you want to group by `event.key` you will pass `group_by_key=True` and no `key`

gtopper replied:
Ah. We dealt with this elsewhere using '$key'. What about `group_by_key=False` and `key_field` being defined?

Michaelliv replied:
> Ah. We dealt with this elsewhere using '$key'. What about **`group_by_key=False` and `key_field`** being defined?

`_Batching` will run as if key_field is not defined.

But i do like the `$key` approach. The only downfall for this is that you have to read the documentation to understand how to use it, whereas switching to `group_by_key=True` will automatically group by `Event.key`. So i guess it's a taste thing. Let me know what is closer to the semantics of the rest of the code base and i'll change it accordingly.

Michaelliv committed:
PR review changes

Michaelliv committed:
Change key extraction mechanism to use '$key' as indication for using Event.key as grouping key.
Added relevant test.
Fixed issue where size of batch dictionary was changing while being iterated.

talIguaz commented on storey/flow.py:
```diff
@@ -615,31 +615,38 @@ def __init__(
         if self._timeout_secs is not None and self._timeout_secs <= 0:
             raise ValueError('Batch timeout cannot be 0 or negative')
 
+        self._extract_key: Optional[Callable[[Event], str]] = None
```
Semantically better to set `self._extract_key = create_extact_key_func()` or something and have that method return lambda rather than initializing it here to None and then init it right after in another place

talIguaz commented on storey/flow.py:
```diff
@@ -716,8 +714,7 @@ class Batch(_Batching):
     :param timeout_secs: Maximum number of seconds to wait before a batch is emitted.
-    :param group_by_key: If true, Batch will group events by given key
-    :param key: The key by which events are grouped, when None will use Event.key for grouping.
+    :param key: The key by which events are grouped, use '$key' to group events by the Event.key property.
```
Specify default behavior

talIguaz commented on tests/test_flow.py:
```diff
@@ -644,6 +645,96 @@ def append_body_and_return(lst, x):
+    values_4 = [i for i in range(4)]
+
+    for _ in range(4):
+        rand_val_1 = choice(values_1)
```
If everything is random you can't really validate specific flows, maybe better to have test specific flows with a specific predefined input

talIguaz replied:
Also better add tests for other scenarios like:
* timeout of a some keys.
* reaching max events of key2 while key1 timeout is "sleeping"
* and a bunch of other variations

Michaelliv committed:
Extended documentation of Batch

Michaelliv committed:
Changed weird None-then-init sequence in _Batch __init__ method, now _create_key_extractor returns a key extration lambda.

gtopper commented on storey/flow.py:
```diff
@@ -608,73 +611,115 @@ async def _handle_completed(self, event, response):
+    Other options may be:
+    Set a '$key' to group events by the Event.key property.
+    set a 'str' key to group events by Event.body[str].
+    set a Callable[Union[Event, Any], Any] to group events by a a custom key extractor.
```
Isn't `Union[Event, Any]` just `Any`? :)

Michaelliv replied:
Indeed something or any == Any, but in case of `full_event=True` its specifically `Event`, you think Any is better?

gtopper commented on storey/flow.py:
```diff
@@ -608,73 +611,115 @@ async def _handle_completed(self, event, response):
+        elif callable(key):
+            return key
+        elif isinstance(key, str):
+            if key == "$key":
```
'$key'

Michaelliv committed:
changed double quotes to single quotes

Michaelliv committed:
Changed type in Batch docs
